src/coreclr/jit/importer.cpp

```diff
@@ -4615,15 +4615,22 @@ GenTree* Compiler::impMathIntrinsic(CORINFO_METHOD_HANDLE method,
     if (!IsIntrinsicImplementedByUserCall(intrinsicName))
 #endif
     {
+        CORINFO_CLASS_HANDLE    tmpClass;
+        CORINFO_ARG_LIST_HANDLE arg;
+        var_types               op1Type;
+        var_types               op2Type;
+
         switch (sig->numArgs)
         {
             case 1:
                 op1 = impPopStack().val;
 
-                assert(varTypeIsFloating(op1));
+                arg     = sig->args;
+                op1Type = JITtype2varType(strip(info.compCompHnd->getArgType(sig, arg, &tmpClass)));
 
-                if (op1->TypeGet() != callType)
+                if (op1->TypeGet() != genActualType(op1Type))
```

Comment:
I really would have preferred to just delete these checks altogether. However, it looks  like there are other places in the importer that are doing similar and that imply the JIT supports implicitly passing `float` args where `double` is expected and that it is supposed to do the right thing.

For example: 
* https://github.com/dotnet/runtime/blob/master/src/coreclr/jit/importer.cpp#L955-L963
* https://github.com/dotnet/runtime/blob/master/src/coreclr/jit/importer.cpp#L11834-L11841
* https://github.com/dotnet/runtime/blob/master/src/coreclr/jit/importer.cpp#L12757-L12771
* https://github.com/dotnet/runtime/blob/master/src/coreclr/jit/importer.cpp#L13204-L13227
* https://github.com/dotnet/runtime/blob/master/src/coreclr/jit/importer.cpp#L17944-L17956

I'm not sure how consistently we are actually doing this. For example, I don't think any of the HWIntrinsic code is attempting to handle this scenario and I'm not sure what else in the JIT might generate such nodes.

Reply:
I can somewhat understand implicitly allowing `float->double` since it is lossless; but the implicit from `double->float` at least seems like it would have been undesirable since that means there is loss of data.

I'm guessing that is has something to do with only having a single spec'd stack type for floats (even though we internally track `TYP_FLOAT` and `TYP_DOUBLE`)